### ISSUE TITLE
MobileAds.Initialize() throws an exception

### ISSUE THREAD

**Issue by ComaneanuMugurel** [googleads/googleads-mobile-unity#854]
Hi,

I'm using admob for almost a year now, but now and then keeps breaking for no obvious reson. It has been working on the app that i am developing, but at some point in the last build it started throwing null exception without me doing anything to the code or the plugin. 

This is the error that is giving me when i try to run this line of code: 
-my line of code: -"MobileAds.Initialize("ca-app-pub-3940256099942544~3347511713");"
-error: - "AndroidPlayer(ADB@127.0.0.1:34999)</i> ADMOB -> Eror when initialize mobile ads: The type initializer for 'GoogleMobileAds.Api.MobileAds' threw an exception."

I tried deleting the current sdk version implemented in the project, and using the latested one, that did not work, keeps doing the same thing. 

Could this be something from the android manifest file? If so, let me know and i will attach the file here.

Also, i've seen that in the documentation, in the example, the banner does not need "MobileAds.Initiliaze" to work, and indeed before this error it did work, but right now when i try to initialize the banner with or without the respective line of code, this error pops up
"Object reference not set to an instance of an object".

Please give a solution to this, there were countless of times admob did not work all of a sudden, and then worked again for no obvoius reason. What should i look for?

**Comment by AsifNaeem1** [comment]
I am facing same problem and working on that issue from past 30 hours. Still unable to resolve that issue. Did you find any solution? If yes, please share

**Comment by gaspetcan** [comment]
Sorry but i cant found it yet

Android için Outlook<https://aka.ms/ghei36>'u edinin

________________________________

**Comment by stowy** [comment]
Hi, thanks for your query.

For support debugging your apps, can we request that you make a request on our [support forum](https://groups.google.com/forum/#!categories/google-admob-ads-sdk/game-engines) at https://groups.google.com/forum/#!categories/google-admob-ads-sdk/game-engines - they will be able to help debug your issue with you efficiently.

When posting to the forum, please also provide specific details including any errors being seen and code showing how to reproduce the problem.

**Comment by harsh2327** [comment]
Found a solution for this.

It looks like Unity was stripping the code out for Reflection. Documented here - https://docs.unity3d.com/Manual/ManagedCodeStripping.html

To solve this issue, one can either
1. Add the following lines in GoogleMobileAdsClientFactory.cs

using UnityEngine.Scripting;
[assembly: Preserve]

OR 

2. Change the "Managed Stripping Level" to Low or less.

**Comment by AsifNaeem1** [comment]
Hello @harsh2327  I also manage that problem by disabling the stripping level. Thank you for you response. 
That was panic for me. I spend over 2 days, I delete each and every file, replace with old setting, then I found the problem is somewhere in player setting. After working so hard I found that stripping level is the reason for MobileAds.Initialize() error.

**Comment by Kendja** [comment]
Since soon we all need to support ARM64(to upload updates on play store), and it requires IL2CPP - managed stripping level can't be disabled anymore.

Solution with "[assembly: Preserve]" from @harsh2327 didn't help unfortunately. 

To make it work for me I needed to remove Reflections from 3 files and call methods directly: 
- MobileAds.cs
- InterstitialAd.cs
- RewardBasedVideoAd.cs

Changes:
...
**MobileAds.cs:**
```
private static IMobileAdsClient GetMobileAdsClient()
{
  return GoogleMobileAdsClientFactory.MobileAdsInstance();
}
```
**InterstitialAd.cs:**
```
this.client = GoogleMobileAdsClientFactory.BuildInterstitialClient();
client.CreateInterstitialAd(adUnitId);
```

**RewardBasedVideoAd.cs:**
```
client = GoogleMobileAdsClientFactory.BuildRewardBasedVideoAdClient();
client.CreateRewardBasedVideoAd();
```

**Comment by stowy** [comment]
Hi thanks for this suggestion, can you please try release 3.18.1 (released today) which no longer has reflection included in the above locations and elsewhere.

**Comment by Kendja** [comment]
@stowy awesome! Thanks! 
I'll check and will let you know

**Comment by themoeedahmed** [comment]
**Make Minify Release from Proguard or Gradle to NONE inside Publishing Settings**

After making all suggested changes.. ads were still not working infect my app gets crash after few seconds of calling MobileAds.Initialize(). I made a change in Build Settings -> Publishing Settings -> Minify -> Open dropdown menu and select None. 

I am glad now my ads are displaying smoothly without being crashing! I really can't even imagine that it really have a direct affect on Admob Plugin Initialization but actually it sorted my issue after 2 days of struggle. I am using Unity 2018.3.0f2 and Admob Plugin!

**Comment by UragayalaDeepika** [comment]
Closing this issue. Feel free to comment here so that we can reopen this issue for you if anyone still have this issue so that we will take a look.

**Comment by diyapkaba** [comment]
I am still having the same problem. The issue is closed but I would like to ask in case someone may help.
I can see test ads while running code on unity editor. When I upload the apk to android, I got this exception:

The type initializer for "GoogleMobileAds.Android.MobileAdsClient" threw an exception.

I checked for possible solutions proposed above. 
- I have no GoogleMobileAdsClientFactory.cs file in the project. 
- Changing Managed Stripping Level or disabling it at all didn't help.

Anyone have any solution? I am using unity 2021.3
